### PR TITLE
feat: add Read Only Mode toggle (web UI + addon config)

### PR DESCRIPTION
**Pull request by kingpanther13**
## What does this PR do?

Adds a server-wide **Read Only Mode**, as discussed in #1569 and originally requested in #549.

When the toggle is on:

- **Catalog**: write-capable tools (`readOnlyHint != true`, fail-closed for unannotated tools) are hidden from the MCP tool list by a list-time transform that reads the live flag.
- **Enforcement**: every write operation is blocked at call time by a new always-installed `ReadOnlyMiddleware`, returning a structured `READ_ONLY_MODE` error that tells the model what was blocked, that no changes were made, and that the user must turn the toggle off to allow writes.
- **Mixed read/write tools** whose read functionality has no pure-read duplicate stay available, with write actions blocked per call:

| Tool | Allowed while Read Only Mode is on |
|---|---|
| `ha_manage_backup` | `scope='edits'` + `action='list'`/`'view'` |
| `ha_manage_addon` | HTTP GET proxy reads (`slug` + `path`, `method='GET'`) |
| `ha_manage_energy_prefs` | `mode='get'` and `dry_run=true` previews (validate/simulate without saving) |
| `ha_manage_pipeline` | `action='list'`/`'get'` |
| `ha_manage_custom_tool` | `list_saved=true` |

- **Proxy dispatch**: calls through `ha_call_write_tool`/`ha_call_delete_tool`/`ha_call_read_tool` are unwrapped and judged on the inner tool, so a proxied write returns the explanatory error instead of a generic not-found (the proxies' category caches no longer contain hidden tools).
- **LLM visibility**: `ha_get_overview` gains projection-proof `read_only_mode` / `read_only_mode_hint` fields, startup server instructions mention the mode, and the structured error covers every remaining path.

Surfaces:

- **Tools tab** (above the search box): "Read Only Mode" toggle - "Toggles all write tools off, and removes ability for tools to make any write or destructive calls." Write-tool rows render forced-off and locked without rewriting saved states (same non-destructive semantics as the beta master gate), so turning the mode off restores prior selections. Exempt tools show a "write operations blocked" note.
- **Server Settings tab**: same flag via `FEATURE_FLAG_FIELDS` (origin tracking, `feature_flags.json` persistence, addon Supervisor routing all reuse existing plumbing).
- **Addon configuration**: `read_only_mode` option in BOTH flavors (options + schema, shared `start.py` env export, translations, DOCS.md). Web UI and addon Configuration page stay in sync through the same Supervisor `/addons/self/options` round-trip as the other non-beta flags.

Default off. Enforcement reads the live settings singleton per call: standalone HTTP mode applies toggles without a restart; addon and stdio modes apply on restart like every other flag.

A real-catalog e2e test guards the invariant that every `MANDATORY_TOOLS` member is either read-only-annotated or exempt, so a future mandatory write tool cannot silently dead-end in read-only mode.

Tests: unit (exemption rules incl. dry-run preview edge cases, middleware verdicts incl. proxy unwrapping and fail-closed empty-catalog, transform filtering, config flag, settings endpoint payload, JS toggle + forced-off rendering, server wiring) and e2e (`tests/src/e2e/policy/test_readonly_mode.py`: catalog hiding, read paths working, exempt read/write split, direct + proxied write blocking, overview reporting, mandatory-tools invariant). Addon structure tests extended for the new option in both flavors.

A separate `chore` commit applies current ruff formatting across `src/` and `tests/`.

## Type of change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [ ] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

## Checklist
- [x] I have updated documentation if needed

### CODE REVIEW

**Comment by kingpanther13** [comment]
Note to self in case I forget later: need to investigate the "middleware" that this created, it might be easier/cleaner to just use the existing tool security middleware.

File: src/ha_mcp/read_only.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    if scope == "edits" and action in ("list", "view"):
+        return None
+    return f"scope={scope!r}, action={action!r}"
+
+
+def _addon_write(args: dict[str, Any]) -> str | None:
+    action = args.get("action")
+    if action:
+        return f"action={action!r}"
+    for param in ("options", "network", "boot", "auto_update", "watchdog"):
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply src/ha_mcp/read_only.py]: Implemented in 477d85ee — extracted `_ADDON_CONFIG_WRITE_PARAMS` as a module-level constant. The schema-drift guard test added in the same batch pins an independent copy of the list and asserts it equals the constant, so any edit to the write-params set deliberately fails a test until consciously acknowledged.

**Comment by kingpanther13** [comment]
## Implementation Summary

**Classification methodology.** Write-capable is derived from MCP annotations (`readOnlyHint != true`, fail-closed for unannotated tools — same default as the policy handlers and search categorizer). Every one of the ~80 registered tools was audited at the implementation level to find mixed read/write tools whose read surface has no pure-read duplicate; exactly five qualify (`ha_manage_backup`, `ha_manage_addon`, `ha_manage_energy_prefs`, `ha_manage_pipeline`, `ha_manage_custom_tool`), each with an argument-level write predicate. Notable judgment calls: energy `dry_run=true` previews are allowed (every write mode short-circuits before `energy/save_prefs`) with a strict `is True` check because the middleware sees raw pre-validation arguments; `ha_manage_addon`'s WebSocket proxy is blocked outright since the initial message can command mutations.

**Design choices.** Enforcement is a new always-installed middleware plus a list-time catalog transform, both reading the live settings singleton per request — zero cost while off, no restart needed in standalone HTTP mode, and independent of the opt-in policy middleware (a read-only-blocked write never queues an approval). Proxy envelopes are unwrapped — including the JSON-string `arguments` form the proxies tolerate — so blocked writes name the inner tool instead of failing as not-found. The Tools-tab toggle force-disables write-tool rows without rewriting saved states (beta-master semantics), so switching the mode off restores prior selections. A real-catalog e2e test guards the invariant that every mandatory tool is read-safe or exempt.

**Review hardening.** A multi-agent review pass plus the Gemini suggestion produced a second batch: fail-closed classification when the catalog lookup itself fails, JSON-string envelope coercion at every unwrap level, UI surfacing for unknown server state and failed toggle saves, an addon option type warning, a schema-drift guard pinning each exempt predicate's inspected arguments against the real tool signatures (independent manifests, so predicate edits deliberately fail a test), and roughly 870 lines of additional unit, JS-behavior, and e2e coverage.

**Problems encountered.** The search-purity e2e initially failed its own anti-vacuous guard (`ha_search_tools` returns a top-level list, not a keyed envelope) — fixed and the failure output confirmed search surfaces only read-safe and exempt tools. Two CodeQL quality findings (implicit string concatenation in a list, a redundant nested import) and the `ha_get_overview` fields-projection completeness gate were caught in CI and fixed.

**Review by Patch76 — COMMENTED** [review]
The enforcement holds up under a write-leak audit. I traced every branch of `ReadOnlyMiddleware.on_call_tool`, `_classify`, the proxy unwrap, and the five exempt predicates — no fail-open path. The predicates are default-block (only an explicit read allowlist returns `None`); empty-catalog and a raising catalog lookup both block; `is_read_safe` keys on `readOnlyHint is True`; and the malformed-proxy-args pass-through is safe because `categorized_call` validates (`VALIDATION_INVALID_JSON` / `INVALID_PARAMETER`, categorized_search.py:317-335) before the category check and the dispatch at :431. `_addon_write` covers every write param currently on `ha_manage_addon` (action, the five config params, `array_patch`, `websocket`, non-GET method), and `_energy_write`'s strict `dry_run is True` short-circuits before `energy/save_prefs`.

One non-blocking thing worth tightening — the schema-drift guard is one-directional. `TestExemptPredicateSchemaDrift` asserts that the names each predicate inspects exist on the real tool (catching a rename or deletion), but nothing asserts the reverse: that every write-capable param or mode on an exempt tool is classified by its predicate. So adding a new write param to `ha_manage_addon` — or a write mode to `ha_manage_energy_prefs`, an action to `ha_manage_pipeline` — without updating the predicate would classify silently as a read, with no test failing. For a feature whose contract is "no write leaks", that's the higher-risk drift direction. A reverse test per exempt tool (every non-read param in the real signature must be in the predicate's write set or explicitly handled) would close it.

The classification is correct as written; the drift-guard direction is the one piece to firm up before this table becomes the pattern future exempt tools copy.

**Comment by kingpanther13** [comment]
@Patch76 Good catch on the guard direction — implemented in 21fa9c5d. Each exempt tool now pins an exact partition of its real signature: every parameter is either inspected by the predicate or explicitly classified as dispatch-gated/read-safe in a justified manifest, and the test fails on any unclassified or stale parameter. A new write-capable parameter on an exempt tool can no longer classify silently as a read — the failure message walks the maintainer through re-reviewing the classification, which also makes the table a safer pattern for future exempt tools to copy. Thanks for the write-leak audit on the enforcement core as well.

**Review by Patch76 — APPROVED** [review]
Round 2 verified: the delta since my last pass is the reverse-direction drift guard (21fa9c5, c819038) — `test_every_tool_param_is_deliberately_classified` now pins an exact partition of each exempt tool's real signature into inspected vs gated-or-read parameters, with no-overlap, no-unclassified, and no-stale assertions. A new write-capable parameter on an exempt tool can no longer slip in as an implicit read, and the failure message walks the maintainer through re-classifying it. The per-parameter justification comments in `_EXEMPT_GATED_OR_READ_ARGS` double as the audit trail for the classification.

From round 1 (code unchanged since 510ada3): every branch of `ReadOnlyMiddleware.on_call_tool`, `_classify`, the proxy unwrap, and the five exempt predicates traced with no fail-open path — default-block predicates, empty-catalog and raising catalog lookup both block, unannotated tools fail closed.

CI is green across the board and all review threads are resolved. Approving.
